File: sdk-ts/src/transport.ts
/**
 * Transport SDK module — TLS config, E2EE messaging, mesh networking.
 *
 * Trade Secret — InFill Systems, LLC. All rights reserved.
 */

import type {
  Node, TLSConfig, HealingResult, CapabilityScope,
} from './types';
import {
  NodeState,
  TLSVersion,
  DowngradeStatus,
  RateLimitResult,
  SignalType,
  DataCategory,
} from './types';

// ---------------------------------------------------------------------------
// TLS
// ---------------------------------------------------------------------------

interface TLSConfigInternal {
  certPath: string;
  keyPath: string;
  caCertPath: string;
  minVersion: TLSVersion;
  verifyClient: boolean;
  isDeveloperMode: boolean;
}

/**
 * TLS configuration and downgrade detection.
 */
export class TransportTLS {
  private _config: TLSConfigInternal | null = null;

  /**
   * Configure TLS.
   */
  configureTLS(
    mode: 'developer' | 'production' = 'developer',
    certPath: string = '',
    keyPath: string = '',
    caCertPath: string = '',
  ): TLSConfigInternal {
    const isDev = mode === 'developer';
    this._config = {
      certPath,
      keyPath,
      caCertPath,
      minVersion: isDev ? TLSVersion.TLS_1_2 : TLSVersion.TLS_1_3,
      verifyClient: !isDev,
      isDeveloperMode: isDev,
    };
    return this._config;
  }

  /**
   * Detect TLS downgrade attacks from request headers.
   * Use lowercase header keys.
   */
  detectDowngrade(headers: Record<string, string>): DowngradeStatus {
    if (!this._config) {
      throw new Error('TLS not configured. Call configureTLS() first.');
    }

    const proto = headers['x-forwarded-proto'] ?? '';
    const tlsVersion = headers['x-tls-version'] ?? '';

    // Plain HTTP = downgrade
    if (proto === 'http') {
      return DowngradeStatus.DOWNGRADE;
    }

    // TLS version below minimum = downgrade
    if (tlsVersion) {
      const versionNum = parseFloat(tlsVersion);
      const minNum = parseFloat(this._config.minVersion);
      if (versionNum < minNum) {
        return DowngradeStatus.DOWNGRADE;
      }
    }

    // HTTPS with valid TLS version = secure
    if (proto === 'https' || tlsVersion) {
      return DowngradeStatus.SECURE;
    }

    return DowngradeStatus.UNKNOWN;
  }

  /**
   * Check rate limit for a node or IP.
   */
  checkRateLimit(key: string): RateLimitResult {
    // Simplified — production would use a sliding window counter
    return RateLimitResult.ALLOWED;
  }
}

// ---------------------------------------------------------------------------
// Mesh
// ---------------------------------------------------------------------------

interface MeshNode {
  node: Node;
  scope: CapabilityScope | null;
  neighbors: Set<string>;
}

/**
 * Self-healing mesh for node isolation and rerouting.
 */
export class MeshModule {
  private _nodes: Map<string, MeshNode> = new Map();
  private _flags: Map<string, Array<{ sourceId: string; signalType: SignalType }>> = new Map();
  private _consensusThreshold: number = 2;

  /**
   * Register a node in the mesh.
   */
  registerNode(node: Node, scope?: CapabilityScope): void {
    this._nodes.set(node.nodeId.uuid, {
      node,
      scope: scope ?? null,
      neighbors: new Set(),
    });
  }

  /**
   * Add a neighbor connection between two nodes.
   */
  addNeighbor(nodeId: string, neighborId: string): void {
    const meshNode = this._nodes.get(nodeId);
    if (meshNode) {
      meshNode.neighbors.add(neighborId);
    }
  }

  /**
   * Flag a node for suspicious behavior.
   */
  flagNode(sourceId: string, targetId: string, signalType: string, details?: Record<string, unknown>): void {
    if (!this._flags.has(targetId)) {
      this._flags.set(targetId, []);
    }
    this._flags.get(targetId)!.push({
      sourceId,
      signalType: signalType as SignalType,
    });
  }

  /**
   * Check if consensus threshold is reached for a node.
   */
  checkConsensus(nodeId: string): boolean {
    const flags = this._flags.get(nodeId) ?? [];
    // Count unique sources
    const sources = new Set(flags.map((f) => f.sourceId));
    return sources.size >= this._consensusThreshold;
  }

  /**
   * Process pending flags and transition node states.
   * Returns list of node IDs that were quarantined.
   */
  processFlags(): string[] {
    const quarantined: string[] = [];

    for (const [nodeId, flags] of this._flags.entries()) {
      const sources = new Set(flags.map((f) => f.sourceId));
      if (sources.size < this._consensusThreshold) continue;

      const meshNode = this._nodes.get(nodeId);
      if (!meshNode) continue;

      // ACTIVE → SUSPECT → QUARANTINED
      const currentState = meshNode.node.state;
      if (currentState === NodeState.ACTIVE) {
        meshNode.node = { ...meshNode.node, state: NodeState.SUSPECT };
      } else if (currentState === NodeState.SUSPECT) {
        meshNode.node = { ...meshNode.node, state: NodeState.QUARANTINED };
        quarantined.push(nodeId);
      }
    }

    return quarantined;
  }

  /**
   * Begin healing a quarantined node.
   */
  beginHealing(nodeId: string, reason: string = ''): HealingResult {
    const meshNode = this._nodes.get(nodeId);
    if (!meshNode) {
      return { success: false, reason: 'Node not found', newState: NodeState.QUARANTINED };
    }

    if (meshNode.node.state !== NodeState.QUARANTINED) {
      return {
        success: false,
        reason: `Node is in ${meshNode.node.state} state, not quarantined`,
        newState: meshNode.node.state,
      };
    }

    meshNode.node = { ...meshNode.node, state: NodeState.HEALING };
    return { success: true, reason: 'Healing initiated', newState: NodeState.HEALING };
  }

  /**
   * Complete healing and restore a node to active.
   */
  completeHealing(nodeId: string): HealingResult {
    const meshNode = this._nodes.get(nodeId);
    if (!meshNode) {
      return { success: false, reason: 'Node not found', newState: NodeState.QUARANTINED };
    }

    if (meshNode.node.state !== NodeState.HEALING) {
      return {
        success: false,
        reason: `Node is in ${meshNode.node.state} state, not healing`,
        newState: meshNode.node.state,
      };
    }

    meshNode.node = { ...meshNode.node, state: NodeState.ACTIVE };
    this._flags.delete(nodeId); // Clear flags on successful healing
    return { success: true, reason: 'Healing complete', newState: NodeState.ACTIVE };
  }
}

// ---------------------------------------------------------------------------
// Transport Module (combined)
// ---------------------------------------------------------------------------

/**
 * Transport: TLS config, downgrade detection, mesh networking.
 */
export class TransportModule {
  public readonly tls: TransportTLS;
  public readonly mesh: MeshModule;

  constructor() {
    this.tls = new TransportTLS();
    this.mesh = new MeshModule();
  }
}

File: sdk/bedrock_sdk/transport.py
"""
Transport SDK module — TLS config, E2EE messaging, mesh networking.

Wraps bedrock.transport and bedrock.mesh with developer-friendly defaults.

Trade Secret — InFill Systems, LLC. All rights reserved.
"""

from typing import List, Optional

from bedrock.transport.security import (
    TransportLayer, TLSConfig, TLSVersion, DowngradeStatus,
    RateLimitConfig, RateLimitResult,
)
from bedrock.mesh.healing import SelfHealingMesh
from bedrock.mesh.detector import SignalType
from bedrock.identity.node import Node, NodeID, NodeState
from bedrock.identity.capabilities import CapabilityScope, DataCategory


class TransportModule:
    """SDK module for transport and mesh networking.

    Provides a simplified API for:
    - TLS configuration and downgrade detection
    - Rate limiting
    - Self-healing mesh operations (flag nodes, consensus, healing)
    - Mesh routing with capability-scope awareness
    """

    def __init__(self, transport: TransportLayer, mesh: SelfHealingMesh):
        self._transport = transport
        self._mesh = mesh

    def configure_tls(
        self,
        mode: str = "developer",
        cert_path: str = "",
        key_path: str = "",
        ca_cert_path: str = "",
    ) -> TLSConfig:
        """Configure TLS for the transport layer.

        Args:
            mode: "developer" (TLS 1.2 minimum, self-signed) or "production"
                (TLS 1.3 minimum, CA-signed).
            cert_path: Path to server certificate.
            key_path: Path to server private key.
            ca_cert_path: Path to CA certificate (production only).

        Returns:
            The configured TLSConfig.
        """
        min_version = TLSVersion.TLS_1_2 if mode == "developer" else TLSVersion.TLS_1_3
        return self._transport.configure_tls(
            cert_path=cert_path,
            key_path=key_path,
            ca_cert_path=ca_cert_path,
            min_version=min_version,
            verify_client=(mode == "production"),
        )

    def detect_downgrade(self, headers: dict) -> str:
        """Detect TLS downgrade attacks from request headers.

        Checks for:
        - x-forwarded-proto indicating plain HTTP
        - x-tls-version below configured minimum

        Args:
            headers: HTTP request headers (use lowercase keys).

        Returns:
            "secure", "downgrade", or "unknown".
        """
        status = self._transport.detect_downgrade(headers)
        return status.value

    def check_rate_limit(self, key: str) -> str:
        """Check rate limit for a node or IP.

        Args:
            key: Node UUID or IP address.

        Returns:
            "allowed", "throttled", or "blocked".
        """
        result = self._transport.check_rate_limit(key)
        return result.value

    def register_mesh_node(
        self,
        name: str,
        categories: Optional[List[str]] = None,
    ) -> Node:
        """Register a new node in the mesh network.

        Args:
            name: Human-readable node name.
            categories: Data categories this node can handle.

        Returns:
            The newly registered Node.
        """
        node = Node(
            node_id=NodeID.generate(),
            name=name,
            state=NodeState.ACTIVE,
        )
        scope = None
        if categories:
            data_cats = [DataCategory(c) for c in categories]
            scope = CapabilityScope(
                node_id=node.node_id.uuid,
                categories=data_cats,
            )
        self._mesh.register_node(node, scope)
        return node

    def flag_node(
        self,
        source_id: str,
        target_id: str,
        signal_type: str,
        details: Optional[dict] = None,
    ) -> None:
        """Flag a node for suspicious behavior in the mesh.

        When enough observers flag a node (≥ consensus_threshold),
        the mesh initiates isolation procedures.

        Args:
            source_id: UUID of the observing node.
            target_id: UUID of the node being flagged.
            signal_type: Type of attack signal detected.
                One of: "credential_stuffing", "unusual_volume",
                "port_scan", "brute_force", "privilege_escalation",
                "attestation_failure", "silent_node", "man_in_the_middle".
            details: Optional dict with additional context.
        """
        signal = SignalType(signal_type)
        self._mesh.flag_node(source_id, target_id, signal, details=details)

    def check_consensus(self, node_id: str) -> bool:
        """Check if enough observers have flagged a node for isolation.

        Args:
            node_id: UUID of the node to check.

        Returns:
            True if consensus threshold is reached.
        """
        return self._mesh.check_consensus(node_id)

    def process_flags(self) -> List[str]:
        """Process pending flags and transition node states.

        Nodes with sufficient flags transition from ACTIVE to SUSPECT.
        Suspect nodes with additional flags transition to QUARANTINED.

        Returns:
            List of node UUIDs that were quarantined.
        """
        return self._mesh.process_flags()

    def begin_healing(self, node_id: str, reason: str = "") -> dict:
        """Begin healing a quarantined node.

        Transitions QUARANTINED → HEALING. The node must pass
        re-attestation before it can return to ACTIVE.

        Args:
            node_id: UUID of the node to heal.
            reason: Reason for initiating healing.

        Returns:
            Dict with "success", "reason", and "new_state" keys.
        """
        result = self._mesh.begin_healing(node_id, reason=reason)
        return {
            "success": result.success,
            "reason": result.reason,
            "new_state": result.new_state.value,
        }

    def complete_healing(self, node_id: str) -> dict:
        """Complete healing and restore a node to ACTIVE.

        Only works after the healing period has elapsed and no new
        flags have been raised during healing.

        Args:
            node_id: UUID of the node to restore.

        Returns:
            Dict with "success", "reason", and "new_state" keys.
        """
        result = self._mesh.complete_healing(node_id)
        return {
            "success": result.success,
            "reason": result.reason,
            "new_state": result.new_state.value,
        }

    def reroute(
        self,
        source_id: str,
        destination_id: str,
        categories: List[str],
    ) -> List[str]:
        """Find an alternate path that avoids quarantined nodes.

        Args:
            source_id: UUID of the source node.
            destination_id: UUID of the destination node.
            categories: Data categories the path must support.

        Returns:
            Ordered list of node UUIDs forming the path.
        """
        data_categories = [DataCategory(c) for c in categories]
        return self._mesh.reroute(source_id, destination_id, data_categories)